tests/cmdline/cmd_showbc_const.py

```diff
@@ -0,0 +1,58 @@
+# cmdline: -v -v
+from micropython import const
+
+_STR = const("foo")
+_EMPTY_TUPLE = const(())
+_TRUE = const(True)
+_FALSE = const(False)
+_SMALLINT = const(33)
+_ZERO = const(0)
+
+
+# Bytecode generated for these if/while statements should contain no JUMP_IF
+# and no instances of string 'Eliminated'
+
+if _STR or _EMPTY_TUPLE:
+    print("Kept")
+if _STR and _EMPTY_TUPLE:
+    print("Eliminated")
+if _TRUE:
+    print("Kept")
+if _SMALLINT:
+    print("Kept")
+if _ZERO and _SMALLINT:
+    print("Eliminated")
+if _FALSE:
+    print("Eliminated")
+
+while _SMALLINT:
+    print("Kept")
+    break
+while _ZERO:
+    print("Eliminated")
+while _FALSE:
+    print("Eliminated")
+
+# These values are stored in variables, and therefore bytecode will contain JUMP_IF
+
+a = _EMPTY_TUPLE or _STR
+if a == _STR:
+    print("Kept")
+
+b = _SMALLINT and _STR
+if b == _STR:
+    print("Kept")
+
+# The compiler is also unable to optimise these expressions, even though the arguments are const,
+# so these also contain JUMP_IF
+
+if (_EMPTY_TUPLE or _STR) == _STR:
```

Comment:
As an aside, it surprised me that the compiler could fold the left-hand side expression but couldn't evaluate the comparisons for these:

```
80 LOAD_CONST_OBJ \.\+='foo'
82 LOAD_CONST_OBJ \.\+='foo'
84 BINARY_OP 2 __eq__
85 POP_JUMP_IF_FALSE 94
87 LOAD_NAME print
89 LOAD_CONST_STRING 'Kept'
91 CALL_FUNCTION n=1 nkw=0
93 POP_TOP
```

I don't know if it's a very likely optimisation for Python code, though.

Reply:
I did not consider implementing logical relation operators (==, <, etc) in constant folding... not sure how useful they would be in real code.